### DISCUSSION TITLE
Getting settings_source value from Settings

### DISCUSSION THREAD

**Discussion by jlost** [pydantic/pydantic#2854] [Question]
Hi, I've been wondering whether it is possible to use a value in a custom settings_source from the Settings, which has been loaded by a higher-priority settings_source. Consider the following example:

```python
from typing import Dict, Any
from pydantic import FilePath, BaseSettings
import yaml
from pathlib import Path

class ProjectConfig(BaseSettings):
    config_file: FilePath = 'src/config-dev.yml'
    foobar: str

    class Config:
        @staticmethod
        def _yaml_config_settings_source(settings: 'ProjectConfig') -> Dict[str, Any]:
            return yaml.safe_load(Path(settings.config_file).read_text('utf-8'))

        @classmethod
        def customise_sources(cls, init_settings, env_settings, file_secret_settings):
            return (
                init_settings,
                env_settings,
                cls._yaml_config_settings_source,
                file_secret_settings,
            )

ProjectConfig(config_file='src/config.yml')
```

This yields the following error:
```
AttributeError: 'ProjectConfig' object has no attribute 'config_file'
```

The idea is that I would like to get the `config_file` path from a higher-priority source such as `init_settings`, `env_settings`, or just the default value.

**Comment by Fkawala** [comment]
@jlost I have (what I think to be) a similar problem and started a new discussion with a simpler/more isolated example [here](https://github.com/samuelcolvin/pydantic/discussions/2894).

**Reply by mattilyra** [reply]
@Fkawala I don't think your linked question applies here. This question is about intermediate intialisation of the class attributes while calling successive `init_sources` methods. The `init_sources` calls are currently implemented in a list comprehension ([here](https://github.com/samuelcolvin/pydantic/blob/428f1e4d342a11dbb0f9866bab6322ab427aec57/pydantic/env_settings.py#L64)) and each call is passed a reference to the _as of yet_ uninitialised settings object.

This question boils down to initialising the settings object after each call to an `init_sources` object

**Comment by mattilyra** [comment]
@jlost I have the exact same use case. I'd like to use path parameter to load a json file _without_ needing to hard code the path. I'd like to pass the path in as an environment variable and read it in the custom `init_sources` function that will then load/parse the json.

Calling the init methods is handled in `BaseSettings` (([here](https://github.com/samuelcolvin/pydantic/blob/428f1e4d342a11dbb0f9866bab6322ab427aec57/pydantic/env_settings.py#L64)) but each call gets a reference to the uninitialised `self` / `BaseSettings`, not to the partially loaded settings. The problem is that `self` (the settings) doesn't get initialised until after all the init methods are called and their values have been consolidated, in `deep_update`.

`sources` below is the list of init source callables for the settings class.
```python
            return deep_update(*reversed([source(self) for source in sources]))
```
I don't think updating the partially loaded settings would have nasty unpredictable side effects. I'm not sure if it would have a performance penalty. Currently the settings tree is updated once for each call of an init method. That would still be true, but the order of the operations would be different. As far as I can tell it would be exactly the same number of update operations for the settings dictionary. Some overhead would result in extra calls to `deep_update`.

I think something along the lines of the following should work.
```python
    init_dict = {}
    for source in reversed(sources):
        _partial_settings = source(self, init_dict)    # pass in the partially loaded settings
        init_dict = deep_update(init_dict, [_partial_settings])  # update partially loaded settings
    return init_dict
```

**Reply by mattilyra** [reply]
I didn't find anything related to this in the currently open list of issues, so I'll open a new one. I think this would be a good additional feature.
